### PR TITLE
[FLINK-12168][table-planner-blink] Support e2e limit, sortLimit, rank, union in blink batch

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

Support limit and add limit it cases to blink batch.
Support sortLimit and add sortLimit it cases to blink batch.
Support rank and add rank it cases to blink batch.
Support union and add union it cases to blink batch.

## Verifying this change

ut

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecRank.scala [hunk 1]
```diff
@@ -103,4 +111,67 @@ class BatchExecRank(
...
+      tableEnv.getConfig,
+      orderByKeys,
+      orderByKeys.map(inputRowType.getTypeAt),
+      orderByCollation.map(_._1),
+      orderByCollation.map(_._2))
+
+
+    //operator needn't cache data
+    val operator = new RankOperator(
+      partitionByCodeGen.generateRecordComparator("PartitionByComparator"),
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecRank.scala]: Looks like we need a dedicated RecordComparatorCodegen, always codegen through SortCodeGenerator seems too heavy for this.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecSortLimit.scala [hunk 2]
```diff
@@ -92,4 +109,45 @@ class BatchExecSortLimit(
...
+    // generate comparator
+    val generator = new SortCodeGenerator(
+      tableEnv.getConfig, keys, keys.map(types(_)), orders, nullsIsLast)
+
+    // TODO If input is ordered, there is no need to use the heap.
+    val operator = new SortLimitOperator(
+      isGlobal,
+      limitStart,
+      limitEnd,
+      generator.generateRecordComparator("SortLimitComparator"))
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecSortLimit.scala]: ditto

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/sort/SortLimitOperator.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+		genComparator = null;
+
+		// reverse the comparision.
+		heap = new PriorityQueue<>((int) limitEnd, (o1, o2) -> comparator.compare(o2, o1));
+		this.collector = new StreamRecordCollector<>(output);
+	}
+
+	@Override
+	public void processElement(StreamRecord<BaseRow> element) throws Exception {
+
```
Comment (KurtYoung) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/sort/SortLimitOperator.java]: delete blank line

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecRank.scala (hunk 1 above)
Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecRank.scala]: Totally agree.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecRank.scala [hunk 4]
```diff
@@ -103,4 +111,66 @@ class BatchExecRank(
...
+        .asInstanceOf[StreamTransformation[BaseRow]]
+    val outputType = FlinkTypeFactory.toInternalRowType(getRowType)
+    val partitionBySortingKeys = partitionKey.toArray
+    // The collation for the partition-by fields is inessential here, we only use the
+    // comparator to distinguish different groups.
+    // (order[is_asc], null_is_last)
+    val partitionBySortCollation = partitionBySortingKeys.map(_ => (true, true))
+
+    val inputRowType = FlinkTypeFactory.toInternalRowType(getInput.getRowType)
+
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecRank.scala]: too many blank lines

**Pull request merged**